prov/sockets/src/sock_fabric.c

```diff
@@ -503,68 +495,105 @@ void sock_getnodename(char *buf, int buflen)
 
 #if HAVE_GETIFADDRS
 	ret = getifaddrs(&ifaddrs);
-	for(ifa = ifaddrs; ifa != NULL; ifa = ifa->ifa_next) {
-		if (ifa->ifa_addr == NULL || (ifa->ifa_addr->sa_family != AF_INET))
-			continue;
-
-		ret = getnameinfo(ifa->ifa_addr, sizeof(struct sockaddr_in),
-				  buf, buflen, NULL, 0, NI_NUMERICHOST);
-		if (ret == 0) {
-			freeifaddrs(ifaddrs);
-			return;
+	if (!ret) {
+		for (ifa = ifaddrs; ifa != NULL; ifa = ifa->ifa_next) {
+			if (ifa->ifa_addr == NULL || !(ifa->ifa_flags & IFF_UP) ||
+			     (ifa->ifa_addr->sa_family != AF_INET))
+				continue;
+
+			ret = getnameinfo(ifa->ifa_addr, sizeof(struct sockaddr_in),
+				  	  buf, buflen, NULL, 0, NI_NUMERICHOST);
+			if (ret == 0) {
+				freeifaddrs(ifaddrs);
+				return;
+			}
 		}
+		freeifaddrs(ifaddrs);
 	}
-	freeifaddrs(ifaddrs);
 #endif
 	/* no reasonable address found, try loopback */
 	strncpy(buf, "127.0.0.1", buflen);
 }
 
-static int sock_getinfo(uint32_t version, const char *node, const char *service,
-			uint64_t flags, struct fi_info *hints, struct fi_info **info)
+struct slist sock_get_list_of_addr(char *hostname, const char *service,
+				    uint64_t flags, struct fi_info *hints)
 {
-	struct fi_info *cur, *tail;
-	enum fi_ep_type ep_type;
-	char hostname[HOST_NAME_MAX];
 	int ret;
+	struct slist addr_list;
+	struct sock_host_list_entry *addr_entry;
+	struct addrinfo ai, *rai = NULL;
+	struct ifaddrs *ifaddrs, *ifa;
+	slist_init(&addr_list);
+	ret = gethostname(hostname, sizeof(hostname));
+	if (ret == 0) {
+		memset(&ai, 0, sizeof(ai));
+		ai.ai_family = AF_INET;
+		ret = getaddrinfo(hostname, service, &ai, &rai);
+		if (!ret)
+			freeaddrinfo(rai);
+	}
 
-	if (!(flags & FI_SOURCE) && hints && hints->src_addr &&
-	    (hints->src_addrlen != sizeof(struct sockaddr_in)))
-		return -FI_ENODATA;
-
-	if (((!node && !service) || (flags & FI_SOURCE)) &&
-	    hints && hints->dest_addr &&
-	    (hints->dest_addrlen != sizeof(struct sockaddr_in)))
-		return -FI_ENODATA;
-
-	ret = sock_verify_info(hints);
-	if (ret)
-		return ret;
+#if HAVE_GETIFADDRS
+	ret = getifaddrs(&ifaddrs);
+	if (!ret) {
+		for (ifa = ifaddrs; ifa != NULL; ifa = ifa->ifa_next) {
+			if (ifa->ifa_addr == NULL || !(ifa->ifa_flags & IFF_UP) ||
+			     (ifa->ifa_addr->sa_family != AF_INET) ||
+			     !strcmp(ifa->ifa_name, "lo"))
+				continue;
 
-	if (!node && !service && !hints) {
-		flags |= FI_SOURCE;
-		sock_getnodename(hostname, sizeof(hostname));
-		node = hostname;
+			addr_entry = calloc(1, sizeof(struct sock_host_list_entry));
+			ret = getnameinfo(ifa->ifa_addr, sizeof(struct sockaddr_in),
+					  addr_entry->hostname, sizeof(addr_entry->hostname),
+					  NULL, 0, NI_NUMERICHOST);
+			if (ret)
+				continue;
+			slist_insert_tail(&addr_entry->entry, &addr_list);
+		}
+		freeifaddrs(ifaddrs);
 	}
-
-	if (!node && !service && !(flags & FI_SOURCE) && !hints->dest_addr) {
-		sock_getnodename(hostname, sizeof(hostname));
-		node = hostname;
+#endif
```

Comment:
ifdef sprinkled in source code is a pain to maintain.  Please surround entire functions where possible, with all ifdef code co-located in the source file, or moved into its own source file.